CARDS-2790: YE: Re-import patient information daily to update inpatient status

acrowthe opened the pull request:
When an inpatient is imported which has a visit:
- Check if the visit can have active survey invites. If not, halt
- If the survey has sent out any emails, mark the visit as on hold and stop sending future emails
- If the survey has not sent out any emails, delete the visit

Important notes while testing this PR:
- Since your computer's date will be set to the past, you will not be able to connect to the internet during testing
- If a CARDS instance encounters a time in the past (eg. you have a running cards instance and set your computer's time backwards, or have a shut down cards instance, set the time backwards and start up the existing cards instance) then some parts of CARDS continue to work but many parts stop working. If this happens, you will have to delete the cards instance and start over. Some examples of breaking components are linking between subjects and forms, search results (such as the dashboards) not showing new entries, and many background tasks.

Testing
- Update the `SubjectSelector.jsx` code to make it easier to generate the correct patients:
  - Update the `createSubjects` function to generate the node name from the entered name, not a uuid
  - Change the following line 854 from:
```
    let url = (subjectParent || "/Subjects") + "/" + uuidv4();
```
to:
```
    let url = (subjectParent || "/Subjects") + "/" + subjectName;
```
- Prepare cards to run offline:
```
docker pull docker.io/library/ubuntu:24.04
mvn clean install -Pdocker
cd compose-cluster
python3 generate_compose_yaml.py --mssql --cards_project  cards4prems --oak_filesystem --dev_docker_image --composum --adminer --smtps --smtps_test_container --smtps_test_mail_path ~/cards-mail --server_address localhost:8080
```
- Edit the `docker-compose.yml` to add the following environment variable: `- NIGHTLY_NOTIFICATIONS_SCHEDULE=0 * * * * ? *`
- Set the current date to 08/28
- Launch cards `docker-compose build && docker-compose up`
- Create a new patient information form. When creating the patient subject, name the subject 111111. Fill out name, DoB, mrn, consents to being contacted by email = yes, email address
- Create a new visit information form for the new patient. Fill out clinic to Integrated Care, Time to the current computer date, provider, location and status.
- Set the date to 09/12. Create a new patient with MRN 111222 and an integrated care visit and a new patient 222111 with an inpatient visit
- Set the date to 9/19, wait a minute and verify that an initial invitation email was sent for 222111
- Set the date to 9/26, wait a minute for 222111's reminder email
- Set the date to 9/28, wait a minute for 111111's invitation email
- Set the date to 10/3, wait for 222111's second reminder
- Set the date to 10/5, wait for 111111's reminder
- Set the date to 10/12, create a new patient 111333 with an integrated care visit and 222222 with an inpatient visit. During this time, 111111 should get a second reminder
- Set the date to 10/13, wait for 111222's invitation email
- Set the date to 10/19, wait for 222222's invitation
- Set the date to 10/20, wait for 111222's reminder
- Set the date to 10/26, wait for 222222's reminder
- Set the date to 10/27, create a new patient 222333 with an inpatient visit. During this time, 111222 should get a second reminder
- Set the date to 10/30. Import a sample inpatient status table:
- In cards/compose-cluster/mssql, run `python3 ./generate_test_inpatient_status.py -n 10 inpatient_sample.sql`
- Edit the generated `inpatient_sample.sql` so it includes MRNs 111111, 111222, 111333, 222111, 222222 and 222333
- Sign into adminer http://localhost:1435/?mssql=mssql&username=sa&db=master&ns=path&import= (password testPassword_)
- Import the inpatient sample sql file (browse, select the file, execute)
- Run the inpatient status import by navigating to http://localhost:8080/Subjects.importClarity?config=Inpatient%20Status

- Verify that patients 111111 and 222111 are unaffected. Verify that patients 111222 and 222222 have their visit information form updated with status on hold. Verify that patients 111333 and 222333 have their visits deleted but patient information form unchanged.
- Set the date to 11/3. Verify that no second reminder email is sent for 222222

acrowthe committed:
CARDS-2790: YE: Re-import patient information daily to update inpatient status
When an inpatient is imported which has a visit:
- Check if the visit can have active survey invites. If not, halt
- If the survey has sent out any emails, mark the visit as on hold and stop sending future emails
- If the survey has not sent out any emails, delete the visit

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/InpatientStatusProcessor.java:
```diff
@@ -0,0 +1,292 @@
+        }
+    }
+
+    private boolean isSurveyValid(Session session, Node visitInformationForm)
```
We already have the `survey_expiry` answer in `Survey Events`, you could use that.

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/InpatientStatusProcessor.java:
```diff
@@ -0,0 +1,292 @@
+                        break;
+                    }
+                } else if ("/Questionnaires/Survey events".equals(questionnairePath)) {
+                    surveyEventsForm = form;
```
There may be two Survey Events, for Integrated Care one may be expired but one may still be valid. You should look at the most recently created one.

sdumitriu reviewed (changes requested):
Looks good, with the exception of multiple Survey Events per visit.

acrowthe committed:
CARDS-2790: YE: Re-import patient information daily to update inpatient status
- Check for submitted forms before deleting visits. If any submitted forms exist,
don't delete their survey event forms and make the visit on hold instead.
Still delete unsubmitted forms/survey event forms for that visit

- If a form has multiple survey event forms, use the most recent for determining expiry

acrowthe committed:
CARDS-2790: YE: Re-import patient information daily to update inpatient status
Update survey expiry check to use survey events form instead of clinic

sdumitriu commented on modules/data-model/forms/api/src/main/java/io/uhndata/cards/forms/api/FormUtils.java:
```diff
@@ -527,6 +531,15 @@ enum SearchType
+     *         If not status flags are present, an empty set will be returned.
+     *         {@code null} may be returned if the node is not a supported type
+     */
+    Set<String> getStatusFlags(Node node);
```
I'm not too happy that this is here, but it's fine until we actually implement the status flags APIs.

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/FormUtilsImpl.java:
```diff
@@ -744,4 +746,24 @@ private JsonValue serializeInputStream(final InputStream value)
+    public Set<String> getStatusFlags(final Node node)
+    {
+        try {
+            if (this.isForm(node) || !this.isAnswerSection(node) || !this.isAnswer(node)) {
```
I don't understand the `!` conditions, why do we exclude answers and sections? Was that just debug code? And the way it is right now, it accepts any node, because at least one of these two conditions will be true, they can't both be false.

sdumitriu commented on modules/data-model/forms/api/src/main/java/io/uhndata/cards/forms/api/FormUtils.java:
```diff
@@ -527,6 +531,15 @@ enum SearchType
+     * Extract the set of status flags from a form, answer section or answer node.
+     * @param node the form, answer section or answer node to pull the status flags from
+     * @return the set of status flags for a form.
+     *         If not status flags are present, an empty set will be returned.
```
```suggestion
     *         If no status flags are present, an empty set will be returned.
```

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/InpatientStatusProcessor.java:
```diff
@@ -0,0 +1,333 @@
+                    if (surveyEventsForm == null) {
+                        surveyEventsForm = form;
+                    } else if (form.getProperty("jcr:created").getDate().after(
+                        surveyEventsForm.getProperty("jcr:created"))
```
Shouldn't it be like this?

```suggestion
                        surveyEventsForm.getProperty("jcr:created").getDate())
```

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/InpatientStatusProcessor.java:
```diff
@@ -0,0 +1,333 @@
+                        surveyEventsForm = form;
+                    }
+                }
+            }
```
Didn't we just do this above? Can we pass `surveyEventsForm` from `processVisitDetails` to `isSurveyValid`?

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/InpatientStatusProcessor.java:
```diff
@@ -0,0 +1,333 @@
+            Node statusQuestion = session.getNode("/Questionnaires/Visit information/status");
+            Node statusAnswer = this.formUtils.getAnswer(visitInformationForm, statusQuestion);
+            String status = (String) this.formUtils.getValue(statusAnswer);
+            if ("cancelled".equals(status) || "entered-in-error".equals(status)) {
```
Include `on-hold` too, otherwise for as long as the patient is in the hospital, we'll keep modifying the form every day.

sdumitriu commented on prems-resources/backend/src/main/java/io/uhndata/cards/prems/internal/importer/InpatientStatusProcessor.java:
```diff
@@ -0,0 +1,333 @@
+                    // Handle visit information forms later
+                    visitInformationForm = form;
+                } else {
+                    form.remove();
```
It's possible to delete some forms before we realize that the visit shouldn't be deleted. I would do two passes: first loop through all the forms and check if the visit has any submitted forms. If not, delete every form, otherwise mark it as on-hold.

sdumitriu committed:
CARDS-2790: YE: Re-import patient information daily to update inpatient status
Some refactoring and bugfixes

sdumitriu committed:
[cleanup] Add final qualifiers where needed

acrowthe committed:
CARDS-2790: YE: Re-import patient information daily to update inpatient status
Fix submitted forms being updated to on hold
